File: sdk/postgres-sdk/java/src/main/java/com/postgres/AliyunEmbedding.java
package com.postgres;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AliyunEmbedding {
    public static final String DEFAULT_API_URL = "https://dashscope.aliyuncs.com/compatible-mode/v1/embeddings";
    public static final String DEFAULT_MODEL = "text-embedding-v4";
    public static final String DEFAULT_API_KEY = "sk-403ca84daa9740df82ce0a1737ceccdf";
    
    private final String apiKey;
    private final String model;
    private final String apiUrl;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    private Integer dimension;
    
    public AliyunEmbedding() {
        this(DEFAULT_API_KEY, DEFAULT_MODEL, DEFAULT_API_URL);
    }
    
    public AliyunEmbedding(String apiKey) {
        this(apiKey, DEFAULT_MODEL, DEFAULT_API_URL);
    }
    
    public AliyunEmbedding(String apiKey, String model) {
        this(apiKey, model, DEFAULT_API_URL);
    }
    
    public AliyunEmbedding(String apiKey, String model, String apiUrl) {
        this.apiKey = apiKey != null ? apiKey : DEFAULT_API_KEY;
        this.model = model != null ? model : DEFAULT_MODEL;
        this.apiUrl = apiUrl != null ? apiUrl : DEFAULT_API_URL;
        this.httpClient = HttpClient.newHttpClient();
        this.objectMapper = new ObjectMapper();
        System.out.println("初始化阿里云百炼Embedding: 模型=" + this.model);
    }
    
    public double[] embedText(String text) throws IOException, InterruptedException {
        Map<String, Object> requestBody = new HashMap<>();
        requestBody.put("model", model);
        requestBody.put("input", text);
        
        String jsonBody = objectMapper.writeValueAsString(requestBody);
        
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiUrl))
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();
        
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() != 200) {
            throw new RuntimeException("阿里云API调用失败: " + response.statusCode() + ", " + response.body());
        }
        
        Map<String, Object> result = objectMapper.readValue(response.body(), Map.class);
        
        @SuppressWarnings("unchecked")
        List<Map<String, Object>> data = (List<Map<String, Object>>) result.get("data");
        
        if (data == null || data.isEmpty()) {
            throw new RuntimeException("阿里云API返回格式错误: " + result);
        }
        
        @SuppressWarnings("unchecked")
        List<Double> embedding = (List<Double>) data.get(0).get("embedding");
        
        double[] vector = new double[embedding.size()];
        for (int i = 0; i < embedding.size(); i++) {
            vector[i] = embedding.get(i);
        }
        
        if (this.dimension == null) {
            this.dimension = vector.length;
            System.out.println("模型维度: " + this.dimension);
        }
        
        return vector;
    }
    
    public double[][] embedTexts(List<String> texts) throws IOException, InterruptedException {
        double[][] vectors = new double[texts.size()][];
        for (int i = 0; i < texts.size(); i++) {
            vectors[i] = embedText(texts.get(i));
        }
        return vectors;
    }
    
    public int getEmbeddingDimension() throws IOException, InterruptedException {
        if (this.dimension == null) {
            embedText("test");
        }
        return this.dimension;
    }
}

File: sdk/postgres-sdk/go/postgres/aliyun_embedding.go
package postgres

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

type AliyunEmbedding struct {
	APIURL    string
	APIKey    string
	Model     string
	HTTPClient *http.Client
}

type AliyunEmbeddingRequest struct {
	Model string `json:"model"`
	Input string `json:"input"`
}

type AliyunEmbeddingResponse struct {
	Object string `json:"object"`
	Data   []struct {
		Object    string    `json:"object"`
		Index     int       `json:"index"`
		Embedding []float64 `json:"embedding"`
	} `json:"data"`
	Model string `json:"model"`
	Usage struct {
		PromptTokens int `json:"prompt_tokens"`
		TotalTokens  int `json:"total_tokens"`
	} `json:"usage"`
}

const (
	DefaultAliyunAPIURL = "https://dashscope.aliyuncs.com/compatible-mode/v1/embeddings"
	DefaultAliyunModel  = "text-embedding-v4"
	DefaultAliyunAPIKey = "sk-403ca84daa9740df82ce0a1737ceccdf"
)

func NewAliyunEmbedding(apiURL, apiKey, model string) *AliyunEmbedding {
	if apiURL == "" {
		apiURL = DefaultAliyunAPIURL
	}
	if apiKey == "" {
		apiKey = DefaultAliyunAPIKey
	}
	if model == "" {
		model = DefaultAliyunModel
	}

	return &AliyunEmbedding{
		APIURL: apiURL,
		APIKey: apiKey,
		Model:  model,
		HTTPClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

func (a *AliyunEmbedding) EmbedText(text string) ([]float64, error) {
	reqBody := AliyunEmbeddingRequest{
		Model: a.Model,
		Input: text,
	}

	jsonBody, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", a.APIURL, bytes.NewBuffer(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+a.APIKey)

	resp, err := a.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var embResp AliyunEmbeddingResponse
	if err := json.Unmarshal(body, &embResp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	if len(embResp.Data) == 0 {
		return nil, fmt.Errorf("no embedding data in response")
	}

	return embResp.Data[0].Embedding, nil
}
